test/performance/odp_ipsec.c

```diff
@@ -818,32 +843,26 @@ run_measure_one_async(ipsec_args_t *cargs,
 			packets_sent += rc;
 		}
 
-		i = 0;
+		num = dequeue_burst(polled_queue, events, MAX_DEQUEUE_BURST);
+		while (num > 0) {
+			odp_packet_t out_pkts[MAX_DEQUEUE_BURST];
 
-		if (cargs->schedule)
-			ev = odp_schedule(NULL,
-					  ODP_SCHED_NO_WAIT);
-		else
-			ev = odp_queue_deq(out_queue);
-
-		while (ev != ODP_EVENT_INVALID) {
-			pkt_out[i] = odp_ipsec_packet_from_event(ev);
-			check_ipsec_result(pkt_out[i]);
-
-			if (cargs->debug_packets)
-				odp_packet_print_data(pkt_out[i], 0,
-						      odp_packet_len(pkt_out[i]));
-			packets_received++;
-			if (cargs->schedule)
-				ev = odp_schedule(NULL,
-						  ODP_SCHED_NO_WAIT);
-			else
-				ev = odp_queue_deq(out_queue);
-			i++;
-		}
+			for (int n = 0; n < num; n++) {
+				odp_packet_t out_pkt;
+
+				out_pkt = odp_ipsec_packet_from_event(events[n]);
+				check_ipsec_result(out_pkt);
 
-		if (i)
-			odp_packet_free_sp(pkt_out, i);
+				if (cargs->debug_packets)
+					odp_packet_print_data(out_pkt, 0,
+							      odp_packet_len(out_pkt));
+				out_pkts[n] = out_pkt;
+			}
+			packets_received += num;
+			odp_packet_free_sp(out_pkts, num);
```

Comment:
Current code aggregates packets from multiple "dequeue" calls and frees in one shot. Proposed code frees packets from single "dequeue_burst" call.

Also, I didn't understand why length of pkt_out or out_pkt array was changed from max_in_flight to MAX_DEQUEUE_BURST.

Reply:
> If there are multiple packets available in the queue, presumably they can be all got through single dequeue_multi/schedule_multi call.

Actually this is the issue. On MRVL platforms, schedule_multi() calls would only return one packet. And freeing single packet is inefficient compared to freeing in a burst. 

If we make the array large enough to accommodate max_in_flight, this issue can be easily worked around. Since this is IPsec perf application, I guess it's better to minimize costs from non-IPsec paths.

Reply:
> Actually this is the issue. On MRVL platforms, schedule_multi() calls would only return one packet. And freeing single packet is inefficient compared to freeing in a burst.

Ok, changed the code to be more like it was, collecting to-be-freed packets from multiple dequeue_multi operations.